FNV support

mklokocka opened the pull request:
Add things I believe to be required for full FNV support.

Namely:

1) A default profile for FNV including patterns for selecting what to pack, optimize, etc.
2) Disabled mesh optimizations for FNV.
3) Ability to specify what to pack using patterns.
4) Option to create `.override` files for each archive created.

Guekka commented on src/bsa_process.cpp:
```diff
@@ -68,4 +71,20 @@ void write_single_archive(const btu::Path &directory_path,
+        }
+        else
+        {
+            std::ofstream out(override_path);
```
Use `btu::common::write_file`

Guekka commented on src/bsa_process.cpp:
```diff
@@ -68,4 +71,20 @@ void write_single_archive(const btu::Path &directory_path,
+
+    if (make_override)
+    {
+        const auto parent_path   = archive_path.value().parent_path();
+        const auto filename      = archive_path.value().stem().u8string() + u8".override";
+        const auto override_path = parent_path / filename;
```
```suggestion
        const auto override_path = std::move(archive_path).replace_extension(".override");
```

Guekka commented on src/gui/AdvancedBSAModule.cpp:
```diff
@@ -46,6 +46,7 @@ void AdvancedBSAModule::settings_to_ui(const Settings &settings)
     ui_->dontMakeLoaded->setChecked(!profile.bsa_make_dummy_plugins);
     ui_->dontRemoveFiles->setChecked(!profile.bsa_remove_files);
     ui_->dontCompress->setChecked(!profile.bsa_allow_compression);
+    ui_->makeOverrides->setChecked(profile.bsa_make_overrides);
```
We should only show this button when the game supports override, don't you think?

mklokocka replied:
Yep, that definitely makes sense.

Guekka commented on src/settings/profile.hpp:
```diff
@@ -101,7 +101,41 @@ class Profile
+            fnv_lod_textures.tex_optimize = OptimizeType::None;
+
+            auto fnv_ui_textures         = PerFileSettings::make_base(game);
+            fnv_ui_textures.pattern      = Pattern(u8"textures/interface/.*\\.dds", Pattern::Type::Regex);
```
Use a wildcard:
```suggestion
            fnv_ui_textures.pattern      = Pattern(u8"textures/interface/*.dds", Pattern::Type::Wildcard);
```

Guekka commented on src/settings/profile.hpp:
```diff
@@ -101,7 +101,41 @@ class Profile
+            fnv_ui_textures.tex_optimize = OptimizeType::None;
+
+            auto fnv_dds         = PerFileSettings::make_base(game);
+            fnv_dds.pattern      = Pattern(u8".*\\.dds", Pattern::Type::Regex);
```
```suggestion
            fnv_dds.pattern      = Pattern(u8"*.dds", Pattern::Type::Wildcard);
```

Guekka commented on src/settings/profile.hpp:
```diff
@@ -101,7 +101,41 @@ class Profile
+
+            auto fnv_dds         = PerFileSettings::make_base(game);
+            fnv_dds.pattern      = Pattern(u8".*\\.dds", Pattern::Type::Regex);
+            fnv_dds.tex.compress = true;
```
Why? FNV only supports older compression formats that create visual artifacts. Compression should probably not be the default

mklokocka replied:
When approaching the default profile I tried to set it up with what I perceive to be the “golden standard” for FNV mods in the community. Which as of now is to use compressed textures at most 2k in resolution. 

Many retexture mods are hilariously oversized (4k+), sometimes even uncompressed. Interestingly nearly all textures >2k are mipmapped by the engine to <2k so using them has no benefit at all.

Guekka commented on src/settings/profile.hpp:
```diff
@@ -101,7 +101,41 @@ class Profile
+            auto fnv_dds         = PerFileSettings::make_base(game);
+            fnv_dds.pattern      = Pattern(u8".*\\.dds", Pattern::Type::Regex);
+            fnv_dds.tex.compress = true;
+            fnv_dds.tex.resize   = btu::tex::Dimension{2048, 2048};
```
Why? Is there a reason to resize by default?

mklokocka replied:
See above reply, basically anything over 2k doesn’t make sense for the game.

Guekka commented on src/settings/profile.hpp:
```diff
@@ -101,7 +101,41 @@ class Profile
+            fnv_dds.tex.resize   = btu::tex::Dimension{2048, 2048};
+
+            auto fnv_tga         = PerFileSettings::make_base(game);
+            fnv_tga.pattern      = Pattern(u8".*\\.tga", Pattern::Type::Regex);
```
```suggestion
            fnv_tga.pattern      = Pattern(u8"*.tga", Pattern::Type::Wildcard);
```

Guekka commented on src/settings/profile.hpp:
```diff
@@ -101,7 +101,41 @@ class Profile
+            fnv_tga.tex_optimize = OptimizeType::None;
+
+            auto fnv_meshes         = PerFileSettings::make_base(game);
+            fnv_meshes.pattern      = Pattern(u8".*\\.nif", Pattern::Type::Regex);
```
```suggestion
            fnv_meshes.pattern      = Pattern(u8"*.nif", Pattern::Type::Wildcard);
```

Guekka commented on src/settings/profile.hpp:
```diff
@@ -101,7 +101,41 @@ class Profile
+            fnv_meshes.nif_optimize = OptimizeType::None;
+
+            auto fnv_sounds    = PerFileSettings::make_base(game);
+            fnv_sounds.pattern = Pattern(u8".*\\.(ogg|wav)", Pattern::Type::Regex);
```
```suggestion
            fnv_sounds.pattern = Pattern(u8"*.[ow][ga][gv]", Pattern::Type::Wildcard);
```

Guekka commented on src/settings/profile.hpp:
```diff
@@ -101,7 +101,41 @@ class Profile
 
         if (game == btu::Game::FNV)
```
Maybe this should be moved to a separate function

mklokocka replied:
Agreed, I just didn’t figure out a place where it would make the most amount of sense.

Guekka reviewed (commented):
First pass review, I haven't looked at the GUI changes yet
It looks good. Please prefer wildcards over regex when possible, they have better performance

mklokocka commented:
Merged your TODO commit into others (and split out the dependency updates). Notably I changed the interface texture pattern back to regex to be able to catch subdirectories.

Also hidden the override toggle for non-FNV games.

Guekka commented:
> I changed the interface texture pattern back to regex to be able to catch subdirectories

Wildcards can match subdirectories. Maybe my explanation of them was not clear?

Guekka reviewed (approved):
Looks good. I'm just not sure I like the split of the BSA module, but it makes sense

mklokocka commented:
Re wildcards: How? There really ought to be some description for how wildcard patterns work. 😅 I thought they were like unix blobs where `*` only catches stuff between directory separators.

Re BSA module split: I tend to agree it's a bit weird having it split, on the other hand it makes semantic sense, since the "pack" setting is tied to patterns more than being a "general" thing.

Guekka commented:
> Re wildcards: How? There really ought to be some description for how wildcard patterns work. 😅 I thought they were like unix blobs where * only catches stuff between directory separators.

The algorithm is [here](https://github.com/Guekka/bethutil/blob/de993e54a0522ddb4788c4951aec04ca45255e2d/include/btu/common/string.hpp#L284)
It's more like a subset of wildcards, which is, I agree, confusing. So `*` matches anything

mklokocka commented:
> It's more like a subset of wildcards, which is, I agree, confusing. So * matches anything
Just tested it again, and I know what the "problem" is - wildcard patterns do not ignore case, unlike regex patterns. So using your proposed wildcard pattern `textures/interface/*.dds` - `textures/interface/something/tex.dds` is correctly skipped, while `textures/Interface/something/tex.dds` is processed.

Guekka commented:
> wildcard patterns do not ignore case
Thanks for trying. Indeed, the case insensitive flag is missing [here](https://github.com/Guekka/cathedral-assets-optimizer/blob/dev/src/settings/per_file_settings.hpp#L55)

In the meantime, we can keep your regex. Tell me when I can merge

mklokocka commented:
As long as you are happy with the changes we can merge. 👍
